src/modules/package-downloader.ts

```diff
@@ -26,57 +26,57 @@ export class PackageDownloader {
+    private async _cloneRepository(packageDir: string, cloneOpts: string[]): Promise<void> {
         await this.git.clone(this.packageConfig.getPackageRepo(), packageDir, cloneOpts);
     }
 
-    async updateRepository(checkRepoAction: CheckRepoActions, packageDir: string, cloneOpts: string[]): Promise<void> {
+    private async _updateRepository(checkRepoAction: CheckRepoActions): Promise<void> {
         const localRepoExists = await this.git.checkIsRepo(checkRepoAction);
 
         if (localRepoExists) {
             await this.git.fetch(['--force', '--tags', '--prune', '--prune-tags']);
-        } else {
-            await this.git.clone(this.packageConfig.getPackageRepo(), packageDir, cloneOpts);
         }
     }
 
-    async checkoutVersion(): Promise<void> {
-        await this.git.checkout(
-            this.packageConfig.version.startsWith(BRANCH_PREFIX)
-                ? this.packageConfig.version.substring(BRANCH_PREFIX.length)
-                : this.packageConfig.version,
-        );
+    private async _checkoutVersion(version: string): Promise<void> {
+        const branchOrTag = version.startsWith(BRANCH_PREFIX) ? version.substring(BRANCH_PREFIX.length) : version;
+        await this.git.checkout(branchOrTag);
     }
 
-    async downloadPackage(option: { checkVersionOnly: boolean }): Promise<SimpleGit> {
-        let packageDir: string = this.packageConfig.getPackageDirectory();
-        let cloneOpts: string[] = [];
-        let checkRepoAction: CheckRepoActions;
-
-        if (option.checkVersionOnly) {
-            packageDir = pathPosix.join(packageDir, '_cache');
-            cloneOpts.push('--bare');
-            checkRepoAction = CheckRepoActions.BARE;
-        } else {
-            packageDir = pathPosix.join(packageDir, this.packageConfig.version);
-            checkRepoAction = CheckRepoActions.IS_REPO_ROOT;
+    private async _checkForValidRepo(packageDir: string, cloneOpts: string[], checkRepoAction: CheckRepoActions): Promise<void> {
+        let directoryExists = CliFileSystem.existsSync(packageDir);
+        if (directoryExists && !(await this.isValidRepo(packageDir, checkRepoAction))) {
+            CliFileSystem.removeSync(packageDir);
+            directoryExists = false;
         }
 
-        if (!CliFileSystem.existsSync(packageDir)) {
-            await this.cloneRepository(packageDir, cloneOpts);
+        if (!directoryExists) {
+            await this._cloneRepository(packageDir, cloneOpts);
         }
+    }
+
+    public async isValidRepo(packageDir: string, checkRepoAction?: CheckRepoActions): Promise<boolean> {
+        return await simpleGit(packageDir).checkIsRepo(checkRepoAction);
+    }
+
+    public async downloadPackage(option: { checkVersionOnly: boolean }): Promise<SimpleGit> {
+        const basePackageDir = this.packageConfig.getPackageDirectory();
+        const packageDir = option.checkVersionOnly ? join(basePackageDir, '_cache') : this.packageConfig.getPackageDirectoryWithVersion();
```

Comment:
Just a matter of personal preference and readability: Prefer if branches (as before your refactoring) to multiple statements which all contain ternary operators with the same condition.

If you have distinct if branches it becomes a lot more obvious that all of these statements are tied to the same condition. If you forget one of these ternary conditions in a manual refactoring, you will quickly find yourself debugging a non-issue you have introduced yourself ;-)

For high level languages such as TypeScript, this does not make any difference, but for lower level languages (C++, Rust) - a dumb compiler might create multiple conditional jumps which end up hurting performance instead of just having one conditional jump only ;-)

Reply:
agree - done (y)